src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+    video::IGPUImage* const dstImage
+)
+    : regions(copyRegions)
+    , srcBuffer(srcBuffer)
+    , dstImage(dstImage)
+    , srcImageFormat(srcImageFormat)
+    , currentBlockInRow(0u)
+    , currentRowInSlice(0u)
+    , currentSliceInLayer(0u)
+    , currentLayerInRegion(0u)
+    , currentRegion(0u)
+{
+    dstImageFormat = dstImage->getCreationParameters().format;
+    if(srcImageFormat == asset::EF_UNKNOWN)
+        srcImageFormat = dstImageFormat;
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+
+    memcpy(&minImageTransferGranularity, &queueFamilyProps.minImageTransferGranularity, sizeof(VkExtent3D));
+
+    // bufferOffsetAlignment:
+        // [x] If Depth/Stencil -> must be multiple of 4
+        // [ ] If multi-planar -> bufferOffset must be a multiple of the element size of the compatible format for the aspectMask of imagesubresource
+        // [x] If Queue doesn't support GRAPHICS_BIT or COMPUTE_BIT ->  must be multiple of 4
+        // [x] bufferOffset must be a multiple of texel block size in bytes
+    bufferOffsetAlignment = dstImageTexelBlockInfo.getBlockByteSize();
+    if(asset::isDepthOrStencilFormat(dstImageFormat))
+        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
+
+    bool queueSupportsCompute = queueFamilyProps.queueFlags.hasFlags(IPhysicalDevice::EQF_COMPUTE_BIT);
+    bool queueSupportsGraphics = queueFamilyProps.queueFlags.hasFlags(IPhysicalDevice::EQF_GRAPHICS_BIT);
+    if((queueSupportsGraphics || queueSupportsCompute) == false)
+        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
+    // TODO: Need to have a function to get equivalent format of the specific plane of this format (in aspectMask)
+    // if(asset::isPlanarFormat(dstImageFormat->getCreationParameters().format))
+    assert(core::is_alignment(bufferOffsetAlignment));
+        
+    // Queues supporting graphics and/or compute operations must report (1,1,1) in minImageTransferGranularity, meaning that there are no additional restrictions on the granularity of image transfer operations for these queues.
+    // Other queues supporting image transfer operations are only required to support whole mip level transfers, thus minImageTransferGranularity for queues belonging to such queue families may be (0,0,0)
+    canTransferMipLevelsPartially = !(minImageTransferGranularity.width == 0 && minImageTransferGranularity.height == 0 && minImageTransferGranularity.depth == 0);
+}
+
+size_t ImageRegionIterator::getMemoryNeededForRemainingRegions() const
+{
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+    assert(dstImageTexelBlockInfo.getBlockByteSize()>0u);
+    auto texelBlockDim = dstImageTexelBlockInfo.getDimension();
+    size_t memoryNeededForRemainingRegions = 0ull;
+    for (uint32_t i = currentRegion; i < regions.size(); ++i)
+    {
+        memoryNeededForRemainingRegions = core::alignUp(memoryNeededForRemainingRegions, bufferOffsetAlignment);
```

Comment:
you need to make a thing to replace operator `+=` so it does alignup then increment